pkg/vault/contracts/VaultExtension.sol

```diff
@@ -760,21 +768,39 @@ contract VaultExtension is IVaultExtension, VaultCommon, Proxy {
     {
         // Retrieve the mapping of tokens and their balances for the specified pool.
         mapping(uint256 tokenIndex => bytes32 packedTokenBalance) storage poolTokenBalances = _poolTokenBalances[pool];
+        RecoveryLocals memory locals;
 
         // Initialize arrays to store tokens and balances based on the number of tokens in the pool.
         IERC20[] memory tokens = _poolTokens[pool];
-        uint256 numTokens = tokens.length;
+        locals.numTokens = tokens.length;
 
-        uint256[] memory balancesRaw = new uint256[](numTokens);
+        uint256[] memory balancesRaw = new uint256[](locals.numTokens);
         bytes32 packedBalances;
 
-        for (uint256 i = 0; i < numTokens; ++i) {
+        for (uint256 i = 0; i < locals.numTokens; ++i) {
             balancesRaw[i] = poolTokenBalances[i].getBalanceRaw();
         }
 
         amountsOutRaw = BasePoolMath.computeProportionalAmountsOut(balancesRaw, _totalSupply(pool), exactBptAmountIn);
 
-        for (uint256 i = 0; i < numTokens; ++i) {
+        // Normally we expect recovery mode withdrawals to be stand-alone operations. If there is a previous add
+        // operation in this transaction, this might be an attack, so we apply the same guardrail used for regular
+        // proportional withdrawals. To keep things simple, all we do is reduce the `amountsOut`, leaving the "fee"
+        // tokens in the pool.
+        locals.swapFeeAmountsRaw = new uint256[](locals.numTokens);
+        locals.chargeRoundtripFee = _addLiquidityCalled().tGet(pool);
+
+        // Don't make the call to retrieve the fee unless we have to.
+        if (locals.chargeRoundtripFee) {
+            locals.swapFeePercentage = _vault.getStaticSwapFeePercentage(pool);
```

Comment:
You don't need an external call here; we're at the vault :)
```suggestion
            locals.swapFeePercentage = _poolConfigBits[pool].getStaticSwapFeePercentage();
```

Reply:
Wow; this is what happens when I work too late at night and my brain's already fried :)